include/boost/gil/color_convert.hpp

```diff
@@ -136,25 +136,52 @@ struct default_color_converter_impl<rgb_t,gray_t> {
 /// y = (1 - b - k) / (1 - k)
+/// where `1` denotes max value of channel type of destination pixel.
+///
+/// The conversion from RGB to CMYK is based on CMY->CMYK (Version 2)
+/// from the Principles of Digital Image Processing - Fundamental Techniques
+/// by Burger, Wilhelm, Burge, Mark J.
+/// and it is a gross approximation not precise enough for professional work.
+///
 template <>
-struct default_color_converter_impl<rgb_t,cmyk_t> {
+struct default_color_converter_impl<rgb_t, cmyk_t>
+{
     template <typename P1, typename P2>
-    void operator()(const P1& src, P2& dst) const {
+    void operator()(const P1& src, P2& dst) const
+    {
         using T2 = typename channel_type<P2>::type;
-        get_color(dst,cyan_t())    = channel_invert(channel_convert<T2>(get_color(src,red_t())));          // c = 1 - r
-        get_color(dst,magenta_t()) = channel_invert(channel_convert<T2>(get_color(src,green_t())));        // m = 1 - g
-        get_color(dst,yellow_t())  = channel_invert(channel_convert<T2>(get_color(src,blue_t())));         // y = 1 - b
-        get_color(dst,black_t())   = (std::min)(get_color(dst,cyan_t()),
-                                                (std::min)(get_color(dst,magenta_t()),
-                                                           get_color(dst,yellow_t())));   // k = minimum(c, m, y)
-        T2 x = channel_traits<T2>::max_value()-get_color(dst,black_t());                  // x = 1 - k
-        if (x>0.0001f) {
-            float x1 = channel_traits<T2>::max_value()/float(x);
-            get_color(dst,cyan_t())    = (T2)((get_color(dst,cyan_t())    - get_color(dst,black_t()))*x1);                // c = (c - k) / x
-            get_color(dst,magenta_t()) = (T2)((get_color(dst,magenta_t()) - get_color(dst,black_t()))*x1);                // m = (m - k) / x
-            get_color(dst,yellow_t())  = (T2)((get_color(dst,yellow_t())  - get_color(dst,black_t()))*x1);                // y = (y - k) / x
-        } else {
-            get_color(dst,cyan_t())=get_color(dst,magenta_t())=get_color(dst,yellow_t())=0;
+        auto const c = channel_invert(channel_convert<T2>(get_color(src, red_t())));   // c = 1 - r
+        auto const m = channel_invert(channel_convert<T2>(get_color(src, green_t()))); // m = 1 - g
+        auto const y = channel_invert(channel_convert<T2>(get_color(src, blue_t())));  // y = 1 - b
+        auto const k = (std::min)(c, (std::min)(m, y)); // k = minimum(c, m, y)
+
+        // Apply color correction, strengthening, reducing non-zero components by
+        // s = 1 / (1 - k) for k < 1, where 1 denotes T2 max, otherwise s = 1 (literal).
+        T2 const k_max = channel_traits<T2>::max_value();
+        T2 const s_div = k_max - k; // x = 1 - k
+        if (s_div > 0.0001f) // TODO: Where 0.0001 as close-to-zero test comes from?
+        {
+            if (k < k_max && s_div < k_max)
+            {
+                float const s = k_max / float(s_div); // s = 1 / (1 - k)
```

Comment:
It might make sense to choose to divide `s` or `k_max` by `s_div` depending on which is lower, because one will probably be smaller than the other, thus allowing us to at least decrease cases of overflow. Though that might lead to inaccuracies.

Reply:
Yes, that could also affect range of output leading to some regressions.

After the `if-else` restructuring, I don't see clear benefit of the division now as the overflow should not happen at all, I hope. Also, for white colours, if `k=0` e.g. then `s_div == k_max`.